src/git-repository-async.js

```diff
@@ -0,0 +1,870 @@
+        return this.repoPromise.then(repo => this.project.relativize(repo.workdir()))
+      })
+    }
+
+    return this.projectAtRoot
+  }
+
+  // Public: Makes a path relative to the repository's working directory.
+  relativize (_path, workingDirectory) {
+    // Cargo-culted from git-utils. The original implementation also handles
+    // this.openedWorkingDirectory, which is set by git-utils when the
+    // repository is opened. Those branches of the if tree aren't included here
+    // yet, but if we determine we still need that here it should be simple to
+    // port.
+    //
+    // The original implementation also handled null workingDirectory as it
+    // pulled it from a sync function that could return null. We require it
+    // to be passed here.
+    if (!_path || !workingDirectory) {
+      return _path
+    }
+
+    if (process.platform === 'win32') {
+      _path = _path.replace(/\\/g, '/')
+    } else {
+      if (_path[0] !== '/') {
+        return _path
+      }
+    }
+
+    if (!/\/$/.test(workingDirectory)) {
+      workingDirectory = `${workingDirectory}/`
+    }
+
+    if (this.isCaseInsensitive) {
+      const lowerCasePath = _path.toLowerCase()
+
+      workingDirectory = workingDirectory.toLowerCase()
+      if (lowerCasePath.indexOf(workingDirectory) === 0) {
+        return _path.substring(workingDirectory.length)
+      } else {
+        if (lowerCasePath === workingDirectory) {
+          return ''
+        }
+      }
+    }
+
+    return _path
+  }
+
+  // Public: Returns a {Promise} which resolves to whether the given branch
+  // exists.
+  hasBranch (branch) {
+    return this.repoPromise
+      .then(repo => repo.getBranch(branch))
+      .then(branch => branch != null)
+      .catch(_ => false)
+  }
+
+  // Public: Retrieves a shortened version of the HEAD reference value.
+  //
+  // This removes the leading segments of `refs/heads`, `refs/tags`, or
+  // `refs/remotes`.  It also shortens the SHA-1 of a detached `HEAD` to 7
+  // characters.
+  //
+  // * `path` An optional {String} path in the repository to get this information
+  //   for, only needed if the repository contains submodules.
+  //
+  // Returns a {Promise} which resolves to a {String}.
+  getShortHead (_path) {
+    return this.getRepo(_path)
+      .then(repo => repo.getCurrentBranch())
+      .then(branch => branch.shorthand())
+  }
+
+  // Public: Is the given path a submodule in the repository?
+  //
+  // * `path` The {String} path to check.
+  //
+  // Returns a {Promise} that resolves true if the given path is a submodule in
+  // the repository.
+  isSubmodule (_path) {
+    return this.repoPromise
+      .then(repo => repo.openIndex())
+      .then(index => {
+        // TODO: This'll probably be wrong if the submodule doesn't exist in the
+        // index yet? Is that a thing?
+        const entry = index.getByPath(_path)
+        const submoduleMode = 57344 // TODO compose this from libgit2 constants
```

Comment:
This seems like a constant that might make sense to put at the top with the status flags.